### PR TITLE
Fix for bugz 1391878 and 1391338 - for multiple wildcard routes requesting the same subdomain

### PR DESCRIPTION
**Pull request by ramr**
fixes bugz https://bugzilla.redhat.com/show_bug.cgi?id=1391338 and 
bugz https://bugzilla.redhat.com/show_bug.cgi?id=1391878

for multiple wildcard routes requesting the same subdomain for the same subdomain and in the same namespace, reject newer routes with 'HostAlreadyClaimed'.

@knobunc / @liggitt   PTAL  Thx

### CODE REVIEW

File: pkg/router/controller/host_admitter.go [hunk 1]
```diff
@@ -227,8 +231,18 @@ func (p *HostAdmitter) displacedRoutes(newRoute *routeapi.Route) ([]*routeapi.Ro
 	// See if any existing wildcard routes block our domain, or if we displace them
 	for i, route := range p.claimedWildcards[wildcardKey] {
 		if route.Namespace == newRoute.Namespace {
-			// Never displace a route in our namespace
-			continue
+			// If new route is a wildcard, still need to check
+			// if it is older as it can displace newer wildcard
+			// routes in the same namespace.
+			if newRoute.Spec.WildcardPolicy != routeapi.WildcardPolicySubdomain || route.Name == newRoute.Name {
```
Comment (liggitt) [inline pkg/router/controller/host_admitter.go]: Multiple wildcard should be able to coexist, we still have path selectivity

**Comment by liggitt** [comment]
How is this different than multiple routes with the same specific hostname coexisting in a namespace (which we allow because they route to separate paths)

File: pkg/router/controller/host_admitter.go (hunk 1 above)
Comment (liggitt) [inline pkg/router/controller/host_admitter.go]: If path is taken into account here, then rejecting overlapping wildcard routes with identical paths seems fine

Comment (liggitt) [inline pkg/router/controller/host_admitter.go]: Adding a few duplicate host+path and wildcard+path routes into the fuzz tests would be good as well

Comment (liggitt) [inline pkg/router/controller/host_admitter.go]: Sorry, I meant add path-differentiated tests to the fuzz tests

Comment (ramr) [inline pkg/router/controller/host_admitter.go]: ack - yup did change that and added some more tests in.

**Comment by ramr** [comment]
@liggitt  addressed your comments. PTAL Thx
[test]

File: pkg/router/controller/host_admitter_test.go [hunk 2]
```diff
@@ -500,14 +506,17 @@ func TestValidRouteAdmissionFuzzing(t *testing.T) {
...
+		makeRoute("ns2", "r2", "com", "", false, makeTime(1*time.Second)),
+		makeRoute("ns3", "r3", "domain1.com", "", false, makeTime(2*time.Second)),
+		makeRoute("ns4", "r4", "domain2.com", "", false, makeTime(3*time.Second)),
+		makeRoute("ns5", "r5", "foo.domain1.com", "", false, makeTime(4*time.Second)),
+		makeRoute("ns6", "r6", "bar.domain1.com", "", false, makeTime(5*time.Second)),
+		makeRoute("ns7", "r7", "sub.foo.domain1.com", "", true, makeTime(6*time.Second)),
+		makeRoute("ns8", "r8", "sub.bar.domain1.com", "", true, makeTime(7*time.Second)),
+		makeRoute("ns8", "r8", "sub.bar.domain1.com", "/p1", true, makeTime(8*time.Second)),
+		makeRoute("ns8", "r8", "sub.bar.domain1.com", "/p2", true, makeTime(9*time.Second)),
+		makeRoute("ns8", "r8", "sub.bar.domain1.com", "/p1/p2/p3", true, makeTime(10*time.Second)),
```
Comment (liggitt) [inline pkg/router/controller/host_admitter_test.go]: give each route it's own name, and add tests for non-wildcard routes with the same host and different paths:

```
		makeRoute("ns8", "r8", "sub.bar.domain1.com", "", true, makeTime(7*time.Second)),
		makeRoute("ns8", "r9", "sub.bar.domain1.com", "/p1", true, makeTime(8*time.Second)),
		makeRoute("ns8", "r10", "sub.bar.domain1.com", "/p2", true, makeTime(9*time.Second)),
		makeRoute("ns8", "r11", "sub.bar.domain1.com", "/p1/p2/p3", true, makeTime(10*time.Second)),
		makeRoute("ns9", "r12", "sub.bar.domain2.com", "", false, makeTime(11*time.Second)),
		makeRoute("ns9", "r13", "sub.bar.domain2.com", "/p1", false, makeTime(12*time.Second)),
```

File: pkg/router/controller/host_admitter.go [hunk 3]
```diff
@@ -170,6 +170,10 @@ func (p *HostAdmitter) addRoute(route *routeapi.Route) error {
 		p.claimedWildcards.RemoveRoute(wildcardKey, displacedRoute)
 
 		msg := fmt.Sprintf("a route in another namespace holds host %s", displacedRoute.Spec.Host)
```
Comment (liggitt) [inline pkg/router/controller/host_admitter.go]:
```
msg := ""
if route.Namespace == displacedRoute.Namespace {
  if route.Spec.WildcardPolicy == routeapi.WildcardPolicySubdomain {
    msg = fmt.Sprintf("wildcard route %s/%s has host *.%s, blocking %s", route.Namespace, route.Name, wildcardKey, displacedRoute.Spec.Host)
  } else {
    msg = fmt.Sprintf("route %s/%s has host %s, blocking %s", route.Namespace, route.Name, route.Spec.Host, displacedRoute.Spec.Host)
  }
} else {
  fmt.Sprintf("a route in another namespace holds host %s", displacedRoute.Spec.Host)
}
```

File: pkg/router/controller/host_admitter.go [hunk 4]
```diff
@@ -227,8 +231,33 @@ func (p *HostAdmitter) displacedRoutes(newRoute *routeapi.Route) ([]*routeapi.Ro
 	// See if any existing wildcard routes block our domain, or if we displace them
 	for i, route := range p.claimedWildcards[wildcardKey] {
 		if route.Namespace == newRoute.Namespace {
-			// Never displace a route in our namespace
```
Comment (liggitt) [inline pkg/router/controller/host_admitter.go]: I would expect the same name and path inequality `continue`s above in the claimedHosts loop.. it's a little weird to divide responsibilities between here (for wildcards) and unique_host (for non-wildcards)

Comment (ramr) [inline pkg/router/controller/host_admitter.go]: Its a double check but I added it in. Yeah, eventually need to merge the two. This ballooned into more.

**Comment by ramr** [comment]
@liggitt  PTAL - you can check the new commit out (easier to review) and if that works I can squash the commits. Thx
[test]

File: pkg/router/controller/host_admitter.go [hunk 5]
```diff
@@ -155,11 +155,16 @@ func (p *HostAdmitter) SetLastSyncProcessed(processed bool) error {
...
 func (p *HostAdmitter) addRoute(route *routeapi.Route) error {
 	// Find displaced routes (or error if an existing route displaces us)
-	displacedRoutes, err := p.displacedRoutes(route)
+	displacedRoutes, err, ownerNamespace := p.displacedRoutes(route)
 	if err != nil {
 		msg := fmt.Sprintf("a route in another namespace holds host %s", route.Spec.Host)
+		if ownerNamespace == route.Namespace {
+			// Use the full error details if we got bumped by a
+			// route in our namespace.
+			msg = err.Error()
```
Comment (liggitt) [inline pkg/router/controller/host_admitter.go]: nice, I like this

**Comment by liggitt** [comment]
LGTM, squash

**Comment by ramr** [comment]
[test]

**Comment by ramr** [comment]
looks like a new parallel test - re[test]

**Comment by ramr** [comment]
Looks like the flake that showed on another PR: #11777  - re[test]

**Comment by liggitt** [comment]
[merge]

**Pull request merged**